composer/distributed/fsdp2.py

```diff
@@ -3,12 +3,109 @@
+    module: nn.Module,
+    auto_wrap_policy: Callable[[nn.Module], Union[bool, dict, None]],
+) -> tuple[list[nn.Module], dict[nn.Module, dict]]:
+    """Identifies direct children of a module that should be wrapped based on the policy."""
+    candidates = []
+    candidate_kwargs = {}
+    for child in module.children():
+        policy_result = auto_wrap_policy(child)
+        if policy_result is True:
+            candidates.append(child)
+        elif isinstance(policy_result, dict):
+            candidates.append(child)
+            candidate_kwargs[child] = policy_result
+    return candidates, candidate_kwargs
+
+
+def legalize_fsdp_wrap_policy(
+    module: nn.Module,
+    auto_wrap_policy: Callable[[nn.Module], Union[bool, dict, None]],
+) -> None:
+    """Legalizes the FSDP wrap policy by ensuring that no submodule has _fsdp_wrap set to True if if the ancestor has _fsdp_wrap set to False."""
+    assert auto_wrap_policy(module) is False, 'The root module must not be wrapped'
+    for submodule in module.modules():
+        submodule_policy = auto_wrap_policy(submodule)
+        if submodule_policy is True or isinstance(submodule_policy, dict):
+            raise ValueError(
+                f'Submodule {submodule} has _fsdp_wrap set to True even though its ancestor {module} has _fsdp_wrap set to False. '
+                f'This will cause errors with FSDP. Please adjust the auto_wrap_policy accordingly.',
+            )
+
+
+def _recursive_apply_fully_shard(
+    root_module: nn.Module,
+    module: nn.Module,
+    fsdp2_config: FSDP2Config,
+    auto_wrap_policy: Callable[[nn.Module], Union[bool, dict, None]],
+    default_kwargs: dict,
+) -> None:
+    """Recursive helper to apply fully_shard based on policy and legalization."""
+    # 1. Identify direct children candidates for sharding based on the policy
+    child_candidates, child_candidate_kwargs = _find_direct_children_to_wrap(module, auto_wrap_policy)
+
+    # 2. Legalize child candidates
+    standalone_child_candidates: list[nn.Module] = []
+    if child_candidates:
+        # Check for tying among the valid candidates based on the policy
+        standalone_child_candidates, tied_children = get_standalone_and_tied_modules(child_candidates)
+        if tied_children:
+            raise ValueError(
+                f'Detected tied parameters between modules designated for FSDP wrapping within {type(module).__name__}. '
+                f'FSDP cannot wrap modules with tied parameters independently at the same level. '
+                f'Please adjust the auto_wrap_policy to ensure no parameter sharing exists between modules to be sharded.',
+            )
+
+        # Check for tying between candidates and the rest of the model (using root_module)
+        legalize_param_sharing_between_modules(root_module, standalone_child_candidates)
+
+    # 3. Legalize all submodules for the modules where _fsdp_wrap is False (none of the descendants should be candidates for sharding)
```

Comment:
I don't think this is how this works. FSDP1 would apply the auto wrap policy check to all modules (https://github.com/pytorch/pytorch/blob/aa285e6512bae860981ffeb2547b85946fde0027/torch/distributed/fsdp/wrap.py#L247), regardless of what happens in the rest of the tree. so for us, any module that has ._fsdp_wrap==True or fsdp_wrap_fn(module)==True/dict, would get put into its own FSDP block (possibly with children that are also in their own FSDP block. lmk if you think I'm misunderstanding, or happy to talk it through live.

Reply:
Ah I forgot this existed, but this is explicitly true for the first level (https://github.com/mosaicml/composer/blob/635d9a73f870d3dafea328a36689cfcf5a2e9c3e/composer/distributed/dist_strategy.py#L397) within the `ComposerModel`, but not true recursively. Will revisit my review in light of this.